src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -72,6 +77,188 @@ internal class Http2FrameWriter
+                    }
+
+                    flushResult = await WriteDataAsync(stream, buffer, buffer.Length, endStream, producer.WriteHeaders);
+                }
+
+                if (producer.IsTimingWrite)
+                {
+                    _timeoutControl.StopTimingWrite();
+                }
+
+                producer.WriteHeaders = false;
+
+                reader.AdvanceTo(buffer.End);
+
+                if ((readResult.IsCompleted && !hasMoreData) || readResult.IsCanceled)
+                {
+                    await reader.CompleteAsync();
+
+                    producer.CompleteResponse();
+                }
+                // We're not going to schedule this again if there's no remaining window.
+                // When the window update is sent, the producer will be re-queued if needed.
+                else if (hasMoreData)
+                {
+                    // We have no more connection window, put this producer in a queue waiting for it to
+                    // a window update to resume the connection.
+                    if (remainingConnection == 0)
+                    {
+                        // Mark the output as waiting for a window upate to resume writing (there's still data)
+                        producer.MarkWaitingForWindowUpdates(true);
+
+                        lock (_windowUpdateLock)
+                        {
+                            _waitingForMoreWindow.Enqueue(producer);
+                        }
+
+                        // Include waiting for window updates in timing writes
+                        if (_minResponseDataRate != null)
+                        {
+                            producer.IsTimingWrite = true;
+                            _timeoutControl.StartTimingWrite();
+                        }
+                    }
+                    else if (remainingStream > 0)
+                    {
+                        // Move this stream to the back of the queue so we're being fair to the other streams that have data
+                        Schedule(producer);
+                    }
+                    else
+                    {
+                        // Mark the output as waiting for a window upate to resume writing (there's still data)
+                        producer.MarkWaitingForWindowUpdates(true);
+
+                        // Include waiting for window updates in timing writes
+                        if (_minResponseDataRate != null)
+                        {
+                            producer.IsTimingWrite = true;
+                            _timeoutControl.StartTimingWrite();
+                        }
+                    }
+                }
+                else if (reschedule)
+                {
+                    Schedule(producer);
+                }
+            }
+            catch (Exception ex)
+            {
+                _log.LogCritical(ex, "The event loop in connection {ConnectionId} failed unexpectedly", _connectionId);
```

Comment:
We should also abort the connection in this case.

Reply:
Is that what happens today?

Reply:
There isn't a connection-level loop like this today.

The closest equivalent is the stream-level loop in the output producer: https://github.com/dotnet/aspnetcore/blob/e9b2adb2a21cc2b0ae5b8f7e93aa4599d6f847b0/src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs#L456

It doesn't abort, but various pipes and completion sources are completed to indicate it's no longer running.

Reply:
Unfortunately it doesn't do what we think it does.